pkg/ns/ns_linux.go

```diff
@@ -152,6 +156,53 @@ func GetNS(nspath string) (NetNS, error) {
 	return &netNS{file: fd}, nil
 }
 
+// Returns a new empty NetNS.
+func TempNetNS() (NetNS, error) {
+	var tempNS NetNS
+	var err error
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	// Create the new namespace in a new goroutine so that if we later fail
+	// to switch the namespace back to the original one, we can safely
+	// leave the thread locked to die without a risk of the current thread
+	// left lingering with incorrect namespace.
+	go func() {
+		defer wg.Done()
+		runtime.LockOSThread()
+
+		var threadNS NetNS
+		// save a handle to current network namespace
+		threadNS, err = getCurrentNSNoLock()
+		if err != nil {
+			err = fmt.Errorf("failed to open current namespace: %v", err)
+			return
+		}
+		defer threadNS.Close()
+
+		// create the temporary network namespace
+		err = unix.Unshare(unix.CLONE_NEWNET)
+		if err != nil {
+			return
+		}
+
+		// get a handle to the temporary network namespace
+		tempNS, err = getCurrentNSNoLock()
+
+		err2 := threadNS.Set()
```

Comment:
> I don't have not so much time to review whole code, but I'm wondering that when temp network namespace is deleted, created by Unshare()?

> tempNS.Close() just closes the netns handle(i.e. inode) in process, not remove netns. (currently pkg/ns/ns_linux.go is not designed to create/delete network ns, actually. Just manipulate 'already created' netns).

@s1061123 
Just after Unshare(), I get the netns handle and revert back the thread to his original netns, so tempNS has no process. There is no call to delete a netns, as soon as there is no reference to a namespace (either a process  using it or a file handle), it'll be deleted.
https://man7.org/linux/man-pages/man7/namespaces.7.html
```
Opening one of the files in this directory (or a file that is
bind mounted to one of these files) returns a file handle for the
 corresponding namespace of the process specified by pid.  As long
as this file descriptor remains open, the namespace will remain
alive, even if all processes in the namespace terminate.  The
file descriptor can be passed to setns(2).
```

Reply:
> Opening one of the files in this directory (or a file that is
bind mounted to one of these files) returns a file handle for the
 corresponding namespace of the process specified by pid.  As long
as this file descriptor remains open, the namespace will remain
alive, even if all processes in the namespace terminate.  The
file descriptor can be passed to setns(2).

This does only mentions 'namespace is not deleted until namespace is open', not 'once all file descriptor is closed, the namespace is automatically removed'. 

> There is no call to delete a netns, as soon as there is no reference to a namespace (either a process using it or a file handle), it'll be deleted.

So please make it clear, what (which component/code) removes netns.

BTW, please add unit-test to guarantee the netns is removed after the close because resource leak is serious issue so we'd like to prevent it as mush as we could.

Reply:
> This does only mentions 'namespace is not deleted until namespace is open', not 'once all file descriptor is closed, the namespace is automatically removed'.

Here a better part of the same man page
```
Namespace lifetime
       Absent any other factors, a namespace is automatically torn down
       when the last process in the namespace terminates or leaves the
       namespace.  However, there are a number of other factors that may
       pin a namespace into existence even though it has no member
       processes.  These factors include the following:

       •  An open file descriptor or a bind mount exists for the
          corresponding /proc/pid/ns/* file.
```

> So please make it clear, what (which component/code) removes netns.

Will improve the comment before the function

> BTW, please add unit-test to guarantee the netns is removed after the close because resource leak is serious issue so we'd like to prevent it as mush as we could.

There is no syscall to list namespaces, you might be confused by the /var/run/netns folder, this is just a convention for "named netns" and if you just create a new netns it'll not appear magically in /var/run/netns

BTW cni repo uses vishvananda/netns
https://github.com/containernetworking/cni/blob/main/pkg/ns/ns_linux.go#L20

Basic example for a tempNetNS (New + Close)
https://github.com/vishvananda/netns/blob/4c46424d73b556b3ea4bc5a7cec9e7376dcb2a73/README.md?plain=1#L40

New is just Unshare + Get
https://github.com/vishvananda/netns/blob/4c46424d73b556b3ea4bc5a7cec9e7376dcb2a73/netns_linux.go#L42
Close just closes the file descriptor
https://github.com/vishvananda/netns/blob/4c46424d73b556b3ea4bc5a7cec9e7376dcb2a73/nshandle_linux.go#L63

NewNamed() if you are curious
https://github.com/vishvananda/netns/blob/4c46424d73b556b3ea4bc5a7cec9e7376dcb2a73/netns_linux.go#L50

In our code we are a bit safer because we are using os.File (so even if we we forget to call Close golang will do it for us)

Reply:
I stopped to review the code, because it is still unclear to me  and you don't answer my question: what (which component/code) removes netns.  I leave it to other reviwers to review.

Reply:
Calling Close() will remove the netns, more precisely it will let the kernel garbage collect it as this is the only reference we have to it.
We can only have a resource leak if we don't close the open file handle to the namespace, and we use os.File so this can't happen (os.File call close in his finalizer if you forgot to)

Is it clear enough ?

Reply:
(I've added `// Calling Close() let the kernel garbage collect the network namespace.`)

Reply:
@s1061123 are my latest explanations clear enough ?

Reply:
Hi @s1061123,
this is marked a "change requested", but don't see any changes to do

Reply:
> Hi @s1061123,
this is marked a "change requested", but don't see any changes to do

Here is what I mentioned, and as above, I stopped to review the code, because it is still unclear to me , you don't answer my question: what (which component/code) removes netns and I don't have enough time to review. I leave it to other reviwers to review. So please find someone else to review the PR.

---
I don't have not so much time to review whole code, but I'm wondering that when temp network namespace is deleted, created by Unshare()?

tempNS.Close() just closes the netns handle(i.e. inode) in process, not remove netns. (currently pkg/ns/ns_linux.go is not designed to create/delete network ns, actually. Just manipulate 'already created' netns).

Reply:
I've answered multiple times that calling Close() is enough to remove the netns, I've shared documentation to support that and other pieces of go code doing the same thing.
I totally understand if you don't have time, and I'm fine having someone else review (any volunteers ?),
simply if the change requested is "please do not leak ressources / netns", then please remove the change requested status as this code does not leak a netns.

Reply:
But let me leave it for further reviewers (still I am not conclude 'we should have temp network namespace').